### PR TITLE
Implemented Messages

### PR DESCRIPTION
**Pull request by exploshot**
Implemented Messages, based on Digitalnote, but with possibility to see the sender.

You can send a message with: 
```
send_msg <recepient_N> -m "Here could be your message"
```
A self destructing message (because of no fee):
```
send_msg <recpient_N> -m "Here could be your self destructing message. Deleted after 5 Minutes" -ttl 5
```

read messages with: 
```
list_msgs
```

Hope you enjoy :)

### CODE REVIEW

**Comment by speqtr** [comment]
@qwertycoin-org I started review. Will publish comments within an hour.

File: .clang-format [hunk 1]
```diff
@@ -2,38 +2,38 @@
 # Licensed under the GNU General Public License, Version 3.
 # See the file LICENSE from this package for details.
 
-# This is the clang-format configuration style to be used by Qt,
+ # This is the clang-format configuration style to be used by Qt,
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 2]
```diff
@@ -2,38 +2,38 @@
...
 # based on the rules from https://wiki.qt.io/Qt_Coding_Style and
 # https://wiki.qt.io/Coding_Conventions
 
----
+ ---
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 3]
```diff
@@ -2,38 +2,38 @@
...
 # Webkit style was loosely based on the Qt style
 BasedOnStyle: WebKit
 
-Standard: Cpp11
+ Standard: Cpp11
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 4]
```diff
@@ -2,38 +2,38 @@
...
 
-# Column width is limited to 100 in accordance with Qt Coding Style.
+ # Column width is limited to 100 in accordance with Qt Coding Style.
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 5]
```diff
@@ -2,38 +2,38 @@
...
 # https://wiki.qt.io/Qt_Coding_Style
 # Note that this may be changed at some point in the future.
 ColumnLimit: 100
 
-# How much weight do extra characters after the line length limit have.
+ # How much weight do extra characters after the line length limit have.
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 6]
```diff
@@ -2,38 +2,38 @@
...
 # PenaltyExcessCharacter: 4
 
-# Disable reflow of qdoc comments: indentation rules are different.
+ # Disable reflow of qdoc comments: indentation rules are different.
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 7]
```diff
@@ -2,38 +2,38 @@
...
 # Translation comments are also excluded.
 CommentPragmas: "^!|^:"
 
-# We want a space between the type and the star for pointer types.
+ # We want a space between the type and the star for pointer types.
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 8]
```diff
@@ -2,38 +2,38 @@
...
 PointerBindsToType: false
 
-# We use template< without space.
+ # We use template< without space.
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 9]
```diff
@@ -2,38 +2,38 @@
...
 SpaceAfterTemplateKeyword: false
 
-# We want to break before the operators, but not before a '='.
+ # We want to break before the operators, but not before a '='.
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 10]
```diff
@@ -2,38 +2,38 @@
...
 BreakBeforeBinaryOperators: NonAssignment
 
-# Braces are usually attached, but not after functions or class declarations.
+ # Braces are usually attached, but not after functions or class declarations.
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 11]
```diff
@@ -48,35 +48,35 @@ BraceWrapping:
     BeforeElse: false
     IndentBraces: false
 
-# When constructor initializers do not fit on one line, put them each on a new line.
+ # When constructor initializers do not fit on one line, put them each on a new line.
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 12]
```diff
@@ -48,35 +48,35 @@ BraceWrapping:
...
 ConstructorInitializerAllOnOneLineOrOnePerLine: true
 # Indent initializers by 4 spaces
 ConstructorInitializerIndentWidth: 4
 
-# Indent width for line continuations.
+ # Indent width for line continuations.
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 13]
```diff
@@ -48,35 +48,35 @@ BraceWrapping:
...
 ContinuationIndentWidth: 8
 
-# No indentation for namespaces.
+ # No indentation for namespaces.
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 14]
```diff
@@ -48,35 +48,35 @@ BraceWrapping:
...
 NamespaceIndentation: None
 
-# Horizontally align arguments after an open bracket.
+ # Horizontally align arguments after an open bracket.
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 15]
```diff
@@ -48,35 +48,35 @@ BraceWrapping:
...
 # The coding style does not specify the following, but this is what gives
 # results closest to the existing code.
 AlignAfterOpenBracket: true
 AlwaysBreakTemplateDeclarations: true
 
-# Ideally we should also allow less short function in a single line, but
+ # Ideally we should also allow less short function in a single line, but
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 16]
```diff
@@ -48,35 +48,35 @@ BraceWrapping:
...
 # clang-format does not handle that.
 AllowShortFunctionsOnASingleLine: Inline
 
-# The coding style specifies some include order categories, but also tells to
+ # The coding style specifies some include order categories, but also tells to
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 17]
```diff
@@ -48,35 +48,35 @@ BraceWrapping:
...
 # separate categories with an empty line. It does not specify the order within
 # the categories. Since the SortInclude feature of clang-format does not
 # re-order includes separated by empty lines, the feature is not used.
 SortIncludes: false
 
-# macros for which the opening brace stays attached.
+ # macros for which the opening brace stays attached.
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 18]
```diff
@@ -48,35 +48,35 @@ BraceWrapping:
...
 ForEachMacros: [foreach, Q_FOREACH, BOOST_FOREACH, forever, Q_FOREVER, QBENCHMARK, QBENCHMARK_ONCE]
 
-# Break constructor initializers before the colon and after the commas.
-BreakConstructorInitializers: BeforeColon
+ # Break constructor initializers before the colon and after the commas.
```
Comment (speqtr) [inline .clang-format]: Unneeded change. Please revert to original state.

File: .clang-format [hunk 19]
```diff
@@ -48,35 +48,35 @@ BraceWrapping:
...
+BreakConstructorInitializers: BeforeColon
```
Comment (speqtr) [inline .clang-format]: Please, configure your IDE to always add a new line at the end of file. In VSCode it can be done by settings `"editor.formatOnSave": true` (or similar property).

File: .gitignore [hunk 20]
```diff
@@ -87,3 +88,6 @@ $RECYCLE.BIN/
 
 # Windows shortcuts
 *.lnk
+
+.vscode
+.vs
```
Comment (speqtr) [inline .gitignore]: Unneeded change. Please add new line at the end of file and configure your IDE to always add a new line at the end of file. In VSCode it can be done by settings `"editor.formatOnSave": true` (or similar property).

File: .travis.yml [hunk 21]
```diff
@@ -109,4 +109,4 @@ after_success:
             lcov --remove coverage.info "/usr/*" "*Xcode.app*" "*/tests/*" --output-file coverage.info
             lcov --list coverage.info
             bash <(curl -s https://codecov.io/bash) -f coverage.info || echo "Upload to codecov.io FAILED"
-        fi
+        fi
```
Comment (speqtr) [inline .travis.yml]: Unneeded change. Please add new line at the end of file and configure your IDE to always add a new line at the end of file. In VSCode it can be done by settings `"editor.formatOnSave": true` (or similar property).

File: CMakeLists.txt [hunk 22]
```diff
@@ -109,4 +109,4 @@ endif()
 
 if(BUILD_WITH_PACKAGE)
     include(project/CPackConfig)
-endif()
+endif()
```
Comment (speqtr) [inline CMakeLists.txt]: Unneeded change. Please add new line at the end of file and configure your IDE to always add a new line at the end of file. In VSCode it can be done by settings `"editor.formatOnSave": true` (or similar property).

File: cmake/external/codecov.cmake [hunk 23]
```diff
@@ -19,4 +19,4 @@ if(CMAKE_CXX_COMPILER_ID MATCHES "Clang|GNU")
     endif()
 else()
     message(STATUS "Compiler ${CMAKE_CXX_COMPILER_ID} doesn't support code-coverage.")
-endif()
+endif()
```
Comment (speqtr) [inline cmake/external/codecov.cmake]: Unneeded change. Please add new line at the end of file and configure your IDE to always add a new line at the end of file. In VSCode it can be done by settings `"editor.formatOnSave": true` (or similar property).

File: cmake/project/CompilerSettings.cmake [hunk 24]
```diff
@@ -27,4 +27,4 @@ elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Intel")
     # Intel C++
 elseif("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
     # Visual Studio C++
-endif()
+endif()
```
Comment (speqtr) [inline cmake/project/CompilerSettings.cmake]: Unneeded change. Please add new line at the end of file and configure your IDE to always add a new line at the end of file. In VSCode it can be done by settings `"editor.formatOnSave": true` (or similar property).

File: lib/CMakeLists.txt [hunk 25]
```diff
@@ -800,4 +800,4 @@ install(TARGETS
         RUNTIME DESTINATION bin
         LIBRARY DESTINATION lib
         ARCHIVE DESTINATION lib/static
-)
+)
```
Comment (speqtr) [inline lib/CMakeLists.txt]: Unneeded change. Please add new line at the end of file and configure your IDE to always add a new line at the end of file. In VSCode it can be done by settings `"editor.formatOnSave": true` (or similar property).

File: lib/Wallet/WalletRpcServer.cpp [hunk 26]
```diff
@@ -139,6 +139,7 @@ void wallet_rpc_server::processRequest(const CryptoNote::HttpRequest& request, C
             { "stop_wallet"      , makeMemberMethod(&wallet_rpc_server::on_stop_wallet)       },
             { "reset"            , makeMemberMethod(&wallet_rpc_server::on_reset)             },
             { "get_payments"     , makeMemberMethod(&wallet_rpc_server::on_get_payments)      },
+			{ "get_messages"	 , makeMemberMethod(&wallet_rpc_server::on_get_messages)	  },
```
Comment (speqtr) [inline lib/Wallet/WalletRpcServer.cpp]: Indent is too big.
```suggestion
            { "get_messages"	 , makeMemberMethod(&wallet_rpc_server::on_get_messages)	  },
```

Comment (speqtr) [reply lib/Wallet/WalletRpcServer.cpp]: Please, always use spaces for indentation. **Never use tabs!** This line is currently indented with tabs.

File: lib/crypto/chacha8.c [hunk 27]
```diff
@@ -165,3 +165,132 @@ void chacha8(const void* data, size_t length, const uint8_t* key, const uint8_t*
...
+      if (length < 64) {
+        memcpy(ctarget, cipher, length);
+      }
+      return;
+    }
+    length -= 64;
+    cipher += 64;
+    data = (uint8_t*)data + 64;
+  }
+}
```
Comment (speqtr) [inline lib/crypto/chacha8.c]: Please add new line at the end of file and configure your IDE to always add a new line at the end of file. In VSCode it can be done by settings `"editor.formatOnSave": true` (or similar property).

File: lib/crypto/crypto.cpp [hunk 28]
```diff
@@ -86,7 +86,7 @@ namespace Crypto {
   void crypto_ops::generate_deterministic_keys(PublicKey &pub, SecretKey &sec, SecretKey& second) {
     lock_guard<mutex> lock(random_lock);
     ge_p3 point;
-    sec = second;
+	sec = second;
```
Comment (speqtr) [inline lib/crypto/crypto.cpp]: Please, replace tab character with 4 spaces!

File: src/SimpleWallet/SimpleWallet.cpp [hunk 29]
```diff
@@ -149,8 +149,8 @@ bool parseUrlAddress(const std::string& url, std::string& address, uint16_t& por
 
   if (addrEnd != std::string::npos) {
     auto portEnd = url.find('/', addrEnd);
-    port = Common::fromString<uint16_t>(url.substr(
-      addrEnd + 1, portEnd == std::string::npos ? std::string::npos : portEnd - addrEnd - 1));
+	auto len = portEnd == std::string::npos ? std::string::npos : (portEnd - addrEnd - 1);
```
Comment (speqtr) [inline src/SimpleWallet/SimpleWallet.cpp]: Replace tab with spaces.

File: src/SimpleWallet/SimpleWallet.cpp [hunk 30]
```diff
@@ -149,8 +149,8 @@ bool parseUrlAddress(const std::string& url, std::string& address, uint16_t& por
...
+	port = Common::fromString<uint16_t>(url.substr(addrEnd + 1, len));
```
Comment (speqtr) [inline src/SimpleWallet/SimpleWallet.cpp]: Replace tab with spaces.

File: src/SimpleWallet/SimpleWallet.cpp [hunk 31]
```diff
@@ -271,6 +274,10 @@ struct TransferCommand {
           return false;
       }
 
+      bool feeFound = false;
+      bool ttlFound = false;
+
+
```
Comment (speqtr) [inline src/SimpleWallet/SimpleWallet.cpp]: No need for two empty lines. One can be removed.

File: src/SimpleWallet/SimpleWallet.cpp [hunk 32]
```diff
@@ -284,7 +291,8 @@ struct TransferCommand {
               logger(ERROR, BRIGHT_RED) << "payment ID has invalid format: \"" << value << "\", expected 64-character string";
               return false;
             }
-          } else if (arg == "-f") {
+          } 
+		  else if (arg == "-f") {
```
Comment (speqtr) [inline src/SimpleWallet/SimpleWallet.cpp]: What is the purpose of moving `else if (...` to a new line? Please, revert this change.

File: src/SimpleWallet/SimpleWallet.cpp [hunk 33]
```diff
@@ -296,12 +304,31 @@ struct TransferCommand {
                 << (m_node.getLastLocalBlockHeaderInfo().majorVersion < CryptoNote::BLOCK_MAJOR_VERSION_4 ? m_currency.minimumFee() : m_node.getMinimalFee());
               return false;
             }
+          } 
+		  else if (arg == "-m") {
```
Comment (speqtr) [inline src/SimpleWallet/SimpleWallet.cpp]: What is the purpose of moving `else if (...` to a new line? Please, revert this change.

File: src/SimpleWallet/SimpleWallet.cpp [hunk 34]
```diff
@@ -296,12 +304,31 @@ struct TransferCommand {
...
+            messages.emplace_back(value);
+          } 
+		  else if (arg == "-ttl") {
```
Comment (speqtr) [inline src/SimpleWallet/SimpleWallet.cpp]: What is the purpose of moving `else if (...` to a new line? Please, revert this change

File: src/SimpleWallet/SimpleWallet.cpp [hunk 35]
```diff
@@ -296,12 +304,31 @@ struct TransferCommand {
...
+                "enter time from 1 to " << (m_currency.mempoolTxLiveTime() / 60) << " minutes";
+              return false;
+            }
           }
         } else {
           WalletLegacyTransfer destination;
           CryptoNote::TransactionDestinationEntry de;
 #ifndef __ANDROID__
-	  std::string aliasUrl;
+	        std::string aliasUrl;
```
Comment (speqtr) [inline src/SimpleWallet/SimpleWallet.cpp]: Replace tab with spaces.

File: src/SimpleWallet/SimpleWallet.cpp [hunk 36]
```diff
@@ -372,6 +399,154 @@ struct TransferCommand {
 
     return true;
   }
+
+  bool parseMessageArgs(LoggerRef& logger, const std::vector<std::string>& args) {
```
Comment (speqtr) [inline src/SimpleWallet/SimpleWallet.cpp]: Inside this function each tab character must be replaced with 4 spaces. Please, never use tabs for indentation!

File: src/SimpleWallet/SimpleWallet.cpp [hunk 37]
```diff
@@ -561,6 +752,48 @@ void printListTransfersItem(LoggerRef& logger, const WalletLegacyTransaction& tx
   logger(INFO, rowColor) << " "; //just to make logger print one endline
 }
 
+void printListMessagesItem(
```
Comment (speqtr) [inline src/SimpleWallet/SimpleWallet.cpp]: Inside this function each tab character must be replaced with 4 spaces. Please, never use tabs for indentation!

File: src/SimpleWallet/SimpleWallet.cpp [hunk 38]
```diff
@@ -707,13 +940,17 @@ simple_wallet::simple_wallet(System::Dispatcher& dispatcher, const CryptoNote::C
...
   m_consoleHandler.setHandler("outputs", boost::bind(&simple_wallet::show_unlocked_outputs_count, this, _1), "Show the number of unlocked outputs available for a transaction");
   m_consoleHandler.setHandler("bc_height", boost::bind(&simple_wallet::show_blockchain_height, this, _1), "Show blockchain height");
   m_consoleHandler.setHandler("transfer", boost::bind(&simple_wallet::transfer, this, _1),
     "transfer <mixin_count> <addr_1> <amount_1> [<addr_2> <amount_2> ... <addr_N> <amount_N>] [-p payment_id] [-f fee]"
     " - Transfer <amount_1>,... <amount_N> to <address_1>,... <address_N>, respectively. "
     "<mixin_count> is the number of transactions yours is indistinguishable from (from 0 to maximum available)");
+  m_consoleHandler.setHandler("send_msg", boost::bind(&simple_wallet::sendMsg, this, _1),
+    "send_msg <addr_N>  -m \"Your Message \" [-ttl time_to_life] [-f fee] [-a anonym] <anonym> 0 or 1 anonymity on or of,"
+	  " if 0 the receiver will not see the last 16 digits from your Address");
```
Comment (speqtr) [inline src/SimpleWallet/SimpleWallet.cpp]: Use spaces instead of tab.

File: src/SimpleWallet/SimpleWallet.cpp [hunk 39]
```diff
@@ -1993,15 +2212,56 @@ bool simple_wallet::listTransfers(const std::vector<std::string>& args) {
...
 
   if (!haveTransfers) {
-    success_msg_writer() << "No transfers";
+    successMsgWriter() << "No transfers";
   }
 
   return true;
 }
+//----------------------------------------------------------------------------------------------------
+bool simple_wallet::listMessages(const std::vector<std::string>& args) {
```
Comment (speqtr) [inline src/SimpleWallet/SimpleWallet.cpp]: Inside this function each tab character must be replaced with 4 spaces. Please, never use tabs for indentation!

File: src/SimpleWallet/SimpleWallet.cpp [hunk 40]
```diff
@@ -2173,51 +2433,181 @@ bool simple_wallet::transfer(const std::vector<std::string> &args) {
...
+    failMsgWriter() << e.what();
   } catch (...) {
-    fail_msg_writer() << "unknown error";
+    failMsgWriter() << "unknown error";
+  }
+
+  return true;
+}
+//----------------------------------------------------------------------------------------------------
+bool simple_wallet::sendMsg(const std::vector<std::string>& args) {
```
Comment (speqtr) [inline src/SimpleWallet/SimpleWallet.cpp]: Inside this function each tab character must be replaced with 4 spaces. Please, never use tabs for indentation!

File: tests/CMakeLists.txt [hunk 41]
```diff
@@ -459,4 +459,4 @@ set_target_properties(QwertycoinTests_UnitTests PROPERTIES
 add_test(
     NAME UnitTests
     COMMAND unit_tests
-)
+)
```
Comment (speqtr) [inline tests/CMakeLists.txt]: Add new line at the end of file.

**Comment by speqtr** [comment]
@qwertycoin-org I finished my review 🙂

**Comment by speqtr** [comment]
@exploshot Updated PR looks much more cleaner and easier to review. Thank you! 🙂

I will do an additional review tomorrow.

PS: PR contains some more tab characters in source code, but no need to fix them, I'll do it while working on "Qt C++ Code Style" task.

**Comment by speqtr** [comment]
@qwertycoin-org could you please rerun the latest CI build on Travis CI? It looks like it failed for macOS on packaging step which looks like an occasional bug of Travis CI which is not related to source changes that @exploshot pushed recently.

**Comment by speqtr** [comment]
NOTE: I think that at this point of development process we can safely ignore codecov failures, because we didn't enable all tests yet. Once I enable all tests successful codecov results will become mandatory.

**Comment by Xecute0101** [comment]
I like this feature very much and this feature might help us in developing a smart contract. But I think users should only be allowed to use self-destructing messages at this point.

**Comment by exploshot** [comment]
Sorry for the "last minute" commit. Now everything should work fine. @Xecute0101 your wishes are now included. Messages can be turned on/off in the Config, smaller Message Size <= 160 Chars which are also editable in the Config.

**Comment by nnpanda** [comment]
160 Chars like a SMS 💃 
Next Update MMS? 320 Chars
